src/pass/lift_if_then_else.cc

```diff
@@ -0,0 +1,403 @@
+
+namespace tvm {
+namespace ir {
+
+using LifterMap = std::unordered_map<const Node*, std::vector<Stmt>>;
+using VarMap = std::unordered_map<const Node*, std::unordered_set<const Node*>>;
+
+/*
+ * This pass tries to lift IfThenElse stmt out of For loop if condition is loop invariant.
+ * For example, given the following block:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        for (k = 0; k < 5; k++)
+ *            if (likely(i*2 < 4))
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * We first detect all IfThenElse stmt and find the corresponding loop invariant For stmt.
+ * Then we lift IfThenElse stmt by one For stmt each step:
+ *
+ * Step 1:
+ * for (i = 0; i < 3; i++)
+ *     for (j = 0; j < 4; j++)
+ *         if (likely(i*2 < 4))
+ *             for (k = 0; k < 5; k++)
+ *                 A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * Step 2:
+ * for (i = 0; i < 3; i++)
+ *     if (likely(i*2 < 4))
+ *         for (j = 0; j < 4; j++)
+ *             for (k = 0; k < 5; k++)
+ *                 A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * In this pass, we only continue detecting possible lifting chance when visiting For,
+ * IfThenElse or AttrStmt Node. For example, for the following block:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        A[i + j] = A[i + j] - 1
+ *        for (k = 0; k < 5; k++)
+ *            if (likely(i*2 < 4))
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * Only the For with k variable will be considered and the resulting stmt would be:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        A[i + j] = A[i + j] - 1
+ *        if (likely(i*2 < 4))
+ *            for (k = 0; k < 5; k++)
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * This pass doesn't do lifting for consecutive IfThenElse stmt. The following
+ * block won't be optimized:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        for (k = 0; k < 5; k++)
+ *            if (likely(i*2 < 4))
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *            if (likely(j > 2))
+ *                A[i+j+k] = B[i+j+k]
+ *
+ */
+class IfThenElseLifter {
+ public:
+  Stmt VisitAndMutate(const Stmt& stmt) {
+    GenerateInternalData(stmt);
+    return PostOrderMutate(stmt);
+  }
+
+ private:
+  void GenerateInternalData(const Stmt& stmt);
+  Stmt PostOrderMutate(const Stmt& stmt);
+  size_t GetUpdatedFor(const Stmt& for_stmt, const Stmt& if_stmt);
+  Stmt LiftIf(const Stmt& if_stmt);
+
+  LifterMap if2for_map_;
+  LifterMap top_for_var_map_;
+  LifterMap for_tracking_map_;
+  LifterMap for2if_map_;
+  VarMap cond_var_map_;
+  std::vector<Stmt> ordered_for_list_;
+};
+
+// Check whether a given IfThenElse stmt is the first one appearing
+// in a For stmt.
+bool is_first_if(const Stmt& for_stmt, const Stmt& if_stmt) {
+  std::vector<size_t> if_hash_list;
+
+  PostOrderVisit(for_stmt.as<For>()->body, [&](const NodeRef& node) {
+    if (node.as<IfThenElse>()) {
+      if_hash_list.push_back(node.hash());
```

Comment:
safer to use `node.get()`? I think you need strict equality instead of hash equality.